python/rafiki/agent.py

```diff
@@ -0,0 +1,195 @@
+           'kStatus','kStatusRunning','kStatusPaused','kStatusError',
+           'kRequest','kResponse']
+
+for t in types:
+    setattr(MsgType,t,MsgType(t))
+
+#####   NOTE the server currently only can handle request sequentially
+
+app = Flask(__name__)
+top_k_=5
+
+class Agent():
+
+    def __init__(self,port):
+        info_queue = Queue()
+        command_queue = Queue()
+        self.p = Process(target=start, args=(port, info_queue,command_queue))
+        self.p.start()
+        self.info_queue=info_queue
+        self.command_queue=command_queue
+        return
+
+    def pull(self):
+        if not self.command_queue.empty():
+            return self.command_queue.get()
+        return None,None 
+
+    def push(self,msg,value):
+        self.info_queue.put((msg,value))
+        return
+
+    def stop(self):
+        self.p.terminate()
+
+def start(port,info_queue,command_queue):
+    global info_queue_, command_queue_, data_
+    info_queue_=info_queue
+    command_queue_=command_queue
+    data_ = []
+    app.run(host='0.0.0.0', port=port)
+    return
+
+def getDataFromInfoQueue(need_return=False):
+    global info_queue_, data_
+    if not need_return:
+        while not info_queue_.empty():
+            msg,d = info_queue_.get()
+            data_.append(d)
+    else:
+        while True: # loop until get answer
+            while not info_queue_.empty():
+                msg,d = info_queue_.get()
+                if msg.is_info():
+                    data_.append(d)
+                else:
+                    return msg,d
+            time.sleep(0.01)
+
+
+@app.route("/")
+@cross_origin()
+def index():
+    try:
+        req=send_from_directory(os.getcwd(),"index.html", mimetype='text/html')
+    except:
+        traceback.print_exc()
+        return "error"
+    return req
+
+# support two operations for user to monitor the training status
+@app.route('/getAllData')
+@cross_origin()
+def getAllData():
+    global data_
+    try:
+        getDataFromInfoQueue()
+    except:
+        traceback.print_exc()
+        return failure("Internal Error")
+    return success(data_)
+
+
+@app.route('/getTopKData')
+@cross_origin()
+def getTopKData():
+    global data_
+    try:
+        k = int(request.args.get("k", top_k_))
+    except:
+        traceback.print_exc()
+        return failure("k should be integer")
+    try:
+        getDataFromInfoQueue()
+    except:
+        traceback.print_exc()
+        return failure("Internal Error")
+    return success(data_[-k:])
+
+@app.route("/api", methods=['POST'])                                                                  
+@cross_origin()                                                                                           
+def api():
+    global info_queue_,command_queue_ 
+    try:
+        file = request.files['image']
+        filename = secure_filename(file.filename)
+        filepath=os.path.join(os.getcwd(), filename)
+        file.save(filepath)
```

Comment:
need to consider to case for other data types, e.g. audio/video, which are stream data that cannot be serialized like images.